### PR TITLE
Fixing incorrect invalidation of positive cudf scale decimal values

### PR DESCRIPTION
**Pull request by hyperbolic2346**
As outlined in https://github.com/NVIDIA/spark-rapids/issues/6693 with great details and reproduction steps, the latest kernel changes were incorrectly processing cudf positive scale decimals. Note that these are represented by Spark as negative scale decimals.

The issue was incorrectly counting the scaled off zeros as significant digits and resulting in either too many digits saved in the value or an invalid value. There were positive scale tests, but they were both insufficient and incorrect. These have been adjusted.

New test added that mimics the failed Spark test. Edge case tests added to verify the reproduction steps in the bug. Original tests adjusted to expect and produce the proper output.

### CODE REVIEW

**Comment by hyperbolic2346** [comment]
build

File: src/main/cpp/src/cast_string.cu [hunk 1]
```diff
@@ -521,7 +521,8 @@ __global__ void string_to_decimal_kernel(T* out,
     }
 
     auto const significant_preceeding_zeros = decimal_location < 0 ? abs(decimal_location) : 0;
```
Comment (mythrocks) [inline src/main/cpp/src/cast_string.cu:523]:
```suggestion
    auto const significant_preceding_zeros = decimal_location < 0 ? abs(decimal_location) : 0;
```

**Review by mythrocks — APPROVED** [review]
One nitpick, but otherwise solid. Thanks, @hyperbolic2346!

**Comment by hyperbolic2346** [comment]
build

**Review by revans2 — DISMISSED** [review]

File: src/main/cpp/src/cast_string.cu [hunk 2]
```diff
@@ -520,8 +520,9 @@ __global__ void string_to_decimal_kernel(T* out,
       }
     }
 
-    auto const significant_preceeding_zeros = decimal_location < 0 ? abs(decimal_location) : 0;
-    auto const zeros_to_decimal             = std::max(0, decimal_location - total_digits);
+    auto const significant_preceding_zeros = decimal_location < 0 ? abs(decimal_location) : 0;
```
Comment (ttnghia) [inline src/main/cpp/src/cast_string.cu:523]:
```suggestion
    auto const significant_preceding_zeros = decimal_location < 0 ? -decimal_location : 0;
```

**Review by ttnghia — APPROVED** [review]
Just a minor nit.

**Comment by hyperbolic2346** [comment]
build

**Review by abellina — APPROVED** [review]

**Comment by abellina** [comment]
Thanks @hyperbolic2346 !
